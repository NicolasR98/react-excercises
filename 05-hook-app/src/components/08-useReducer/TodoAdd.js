import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {
	const [{ description }, handleInputChange, resetForm] = useForm({
		description: '',
	});

	const handleSubmit = (e) => {
		e.preventDefault();

		if (description.trim().length <= 1) return;

		const newTodo = {
			id: new Date().getTime(),
			desc: description,
			done: false,
		};

		handleAddTodo(newTodo);
		resetForm();
	};

	return (
		<>
			<h4>Add ToDo</h4>
			<hr />

			<form onSubmit={handleSubmit}>
				<input
					onChange={handleInputChange}
					value={description}
					className="form-control"
					type="text"
					name="description"
					placeholder="Do ..."
					autoComplete="off"
				/>
				<button type="submit" className="btn btn-success col-12 mt-3">
					Add ToDo
				</button>
			</form>
		</>
	);
};
