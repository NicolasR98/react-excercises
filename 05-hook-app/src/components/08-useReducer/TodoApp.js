import React, { useEffect, useReducer } from 'react';
import { useForm } from '../../hooks/useForm';
import { todoReducer } from './todoReducer';

const init = () => JSON.parse(localStorage.getItem('todos')) || [];

export const TodoApp = () => {
	const [todos, dispatch] = useReducer(todoReducer, [], init);

	const [{ description }, handleInputChange, resetForm] = useForm({
		description: '',
	});

	useEffect(() => {
		localStorage.setItem('todos', JSON.stringify(todos));
	}, [todos]);

	const handleDelete = (todoId) => {
		const action = {
			type: 'delete',
			payload: todoId,
		};
		dispatch(action);
	};

	const handleToggle = (todoId) => {
		dispatch({
			type: 'toggle',
			payload: todoId,
		});
	};

	const handleSubmit = (e) => {
		e.preventDefault();

		if (description.trim().length <= 1) return;

		const newTodo = {
			id: new Date().getTime(),
			desc: description,
			done: false,
		};

		const action = {
			type: 'add',
			payload: newTodo,
		};

		dispatch(action);
		resetForm();
	};

	return (
		<div>
			<h1>TodoApp ({todos.length})</h1>
			<hr />

			<div className="row">
				<div className="col-7">
					<ul className="list-group list-group-flush">
						{todos.map((todo, index) => (
							<li key={todo?.id} className="list-group-item">
								<p
									className={`${todo.done && 'complete'}`}
									onClick={() => handleToggle(todo.id)}
								>
									{index + 1}. {todo?.desc}
								</p>
								<button
									onClick={() => handleDelete(todo.id)}
									className="btn btn-danger btn-sm"
								>
									Delete
								</button>
							</li>
						))}
					</ul>
				</div>
				<div className="col-5">
					<h4>Add ToDo</h4>
					<hr />

					<form onSubmit={handleSubmit}>
						<input
							onChange={handleInputChange}
							value={description}
							className="form-control"
							type="text"
							name="description"
							placeholder="Do ..."
							autoComplete="off"
						/>
						<button type="submit" className="btn btn-success col-12 mt-3">
							Add ToDo
						</button>
					</form>
				</div>
			</div>
		</div>
	);
};
