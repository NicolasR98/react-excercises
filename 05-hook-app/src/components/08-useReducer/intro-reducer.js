const initialState = [
	{
		id: 1,
		todo: 'buy groceries',
		done: false,
	},
];

const todoReducer = (state = initialState, action) => {
	if (action?.type === 'add') {
		return [...state, action.payload];
	}

	return state;
};

let todos = todoReducer();

const newTodo = {
	id: 1,
	todo: 'buy groceries',
	done: true,
};

const addTodoAction = {
	type: 'add',
	payload: newTodo,
};

todos = todoReducer(todos, addTodoAction);

console.log(todos);
