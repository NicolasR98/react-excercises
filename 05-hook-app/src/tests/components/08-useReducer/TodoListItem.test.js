import React from 'react';
import { shallow } from 'enzyme';
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';
import { demoTodos } from '../../fixtures/demoTodos,';

describe('Tests on <TodoListItem />', () => {
	const [todo] = demoTodos;

	const handleToggle = jest.fn();
	const handleDelete = jest.fn();

	const wrapper = shallow(
		<TodoListItem
			todo={todo}
			index={0}
			handleToggle={handleToggle}
			handleDelete={handleDelete}
		/>
	);

	test('should display correctly', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test("should call the function 'handleDelete'", () => {
		//find and click on 'deleteButton'
		wrapper.find('button').simulate('click');

		expect(handleDelete).toHaveBeenCalledTimes(1);
		expect(handleDelete).toHaveBeenCalledWith(todo?.id);
	});

	test("should call the function 'handleToggle'", () => {
		//find and click on 'todo'
		wrapper.find('p').simulate('click');

		expect(handleToggle).toHaveBeenCalledTimes(1);
		expect(handleToggle).toHaveBeenCalledWith(todo?.id);
	});

	test('should correctly display the text', () => {
		const p = wrapper.find('p');

		expect(p.text().trim()).toBe(`1. ${todo.desc}`);
	});

	test('should have the class `complete` if `TODO.done = true`', () => {
		const newTodo = { ...todo, done: true };

		const wrapper = shallow(
			<TodoListItem
				todo={newTodo}
				index={0}
				handleToggle={handleToggle}
				handleDelete={handleDelete}
			/>
		);

        expect(wrapper.find('p').hasClass('complete')).toBeTruthy();
	});
});
