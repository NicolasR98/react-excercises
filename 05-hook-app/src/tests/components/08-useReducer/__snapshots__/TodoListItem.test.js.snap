// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <TodoListItem /> should display correctly 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className="false"
    onClick={[Function]}
  >
    1
    . 
    Learn react
  </p>
  <button
    className="btn btn-danger btn-sm"
    onClick={[Function]}
  >
    Delete
  </button>
</li>
`;
